planner/core/task.go

```diff
@@ -240,6 +240,72 @@ func (p *PhysicalIndexMergeJoin) GetCost(outerTask, innerTask task) float64 {
 	return outerTask.cost() + innerPlanCost + cpuCost + memoryCost
 }
 
+func (p *PhysicalIndexHashJoin) attach2Task(tasks ...task) task {
+	innerTask := p.innerTask
+	outerTask := finishCopTask(p.ctx, tasks[1-p.InnerChildIdx].copy())
+	if p.InnerChildIdx == 1 {
+		p.SetChildren(outerTask.plan(), innerTask.plan())
+	} else {
+		p.SetChildren(innerTask.plan(), outerTask.plan())
+	}
+	p.schema = buildPhysicalJoinSchema(p.JoinType, p)
+	return &rootTask{
+		p:   p,
+		cst: p.GetCost(outerTask, innerTask),
+	}
+}
+
+// GetCost computes the cost of index merge join operator and its children.
+func (p *PhysicalIndexHashJoin) GetCost(outerTask, innerTask task) float64 {
+	var cpuCost float64
+	outerCnt, innerCnt := outerTask.count(), innerTask.count()
+	// Add the cost of evaluating outer filter, since inner filter of index join
+	// is always empty, we can simply tell whether outer filter is empty using the
+	// summed length of left/right conditions.
+	if len(p.LeftConditions)+len(p.RightConditions) > 0 {
+		cpuCost += CPUFactor * outerCnt
+		outerCnt *= selectionFactor
+	}
+	// Cost of extracting lookup keys.
+	innerCPUCost := CPUFactor * outerCnt
+	// Cost of sorting and removing duplicate lookup keys:
+	// (outerCnt / batchSize) * (batchSize * Log2(batchSize) + batchSize) * CPUFactor
+	batchSize := math.Min(float64(p.ctx.GetSessionVars().IndexJoinBatchSize), outerCnt)
+	if batchSize > 2 {
+		innerCPUCost += outerCnt * (math.Log2(batchSize) + 1) * CPUFactor
+	}
+	// Add cost of building inner executors. CPU cost of building copTasks:
+	// (outerCnt / batchSize) * (batchSize * distinctFactor) * CPUFactor
+	// Since we don't know the number of copTasks built, ignore these network cost now.
+	innerCPUCost += outerCnt * distinctFactor * CPUFactor
+	concurrency := float64(p.ctx.GetSessionVars().IndexLookupJoinConcurrency)
+	cpuCost += innerCPUCost / concurrency
+	// CPU cost of building hash table for outer results concurrently.
+	// (outerCnt / batchSize) * (batchSize * CPUFactor)
+	outerCPUCost := outerCnt * CPUFactor
+	cpuCost += outerCPUCost / concurrency
+	// Cost of probing hash table concurrently.
+	numPairs := outerCnt * innerCnt
+	if p.JoinType == SemiJoin || p.JoinType == AntiSemiJoin ||
+		p.JoinType == LeftOuterSemiJoin || p.JoinType == AntiLeftOuterSemiJoin {
+		if len(p.OtherConditions) > 0 {
+			numPairs *= 0.5
+		} else {
+			numPairs = 0
+		}
+	}
+	probeCost := numPairs * CPUFactor / concurrency
```

Comment:
Could this `probeCost` be directly divided by `concurrency`? If data is very much, the index hash join would not fall back to linear execution?